Fix zip serialization for file > 2GiB

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #40723 Support Pathlike for zipfile serialization
* **#40722 Fix zip serialization for file > 2GiB**

This narrowing conversion caused wrap-around of the value returned by python `readinto`, and thus everything went haywire since miniz thought that we weren't able to read enough data

Differential Revision: [D22294016](https://our.internmc.facebook.com/intern/diff/D22294016)

Commit:
Fix zip serialization for file > 2GiB

[ghstack-poisoned]

zdevito commented on test/test_serialization.py:
```diff
@@ -673,6 +673,15 @@ def test(name_or_buffer):
         test(io.BytesIO())
 
+    # Ensure large zip64 serialization works properly
+    def test_serialization_2gb_file(self):
```
I recall having a test for big zip files but it got disabled for being too slow :(

Commit:
Update on "Fix zip serialization for file > 2GiB"


This narrowing conversion caused wrap-around of the value returned by python `readinto`, and thus everything went haywire since miniz thought that we weren't able to read enough data

Differential Revision: [D22294016](https://our.internmc.facebook.com/intern/diff/D22294016)

[ghstack-poisoned]

Commit:
Update on "Fix zip serialization for file > 2GiB"


This narrowing conversion caused wrap-around of the value returned by python `readinto`, and thus everything went haywire since miniz thought that we weren't able to read enough data

Differential Revision: [D22294016](https://our.internmc.facebook.com/intern/diff/D22294016)

[ghstack-poisoned]

Commit:
Update on "Fix zip serialization for file > 2GiB"


This narrowing conversion caused wrap-around of the value returned by python `readinto`, and thus everything went haywire since miniz thought that we weren't able to read enough data

Differential Revision: [D22294016](https://our.internmc.facebook.com/intern/diff/D22294016)

[ghstack-poisoned]

peterjc123 reviewed (commented):
long is a 32-bit type with MSVC. That is probably why this test fail on Windows.

peterjc123 commented:
@jamesr66a I've put up a fix in https://github.com/pytorch/pytorch/pull/40783. Could you please take a look there?
